ftdc/ftdc.go

```diff
@@ -165,6 +173,70 @@ func (ftdc *FTDC) Remove(name string) {
 		"name", name, "generationId", ftdc.inputGenerationID)
 }
 
+// Start spins off the background goroutine for collecting/writing FTDC data.
+func (ftdc *FTDC) Start() {
+	ftdc.readStatsWorker = utils.NewStoppableWorkerWithTicker(time.Second, ftdc.statsReader)
+	go ftdc.statsWriter()
+}
+
+func (ftdc *FTDC) statsReader(ctx context.Context) {
+	datum := ftdc.constructDatum()
+	if datum.generationID == 0 {
+		// No "statsers" were `Add`ed. No data to write out.
+		return
+	}
+
+	// `Debugw` does not seem to serialize any of the `datum` value.
+	ftdc.logger.Debugf("Metrics collected. Datum: %+v", datum)
+
+	select {
+	case ftdc.datumCh <- datum:
+		break
+	case <-ftdc.outputWorkerDone:
+		break
+	case <-ctx.Done():
+		break
+	}
+}
+
+func (ftdc *FTDC) statsWriter() {
+	defer func() {
+		if ftdc.currOutputFile != nil {
+			ftdc.currOutputFile.Close()
+		}
+		close(ftdc.outputWorkerDone)
+	}()
+
+	datumsWritten := 0
+	for datum := range ftdc.datumCh {
+		if err := ftdc.writeDatum(datum); err != nil {
+			ftdc.logger.Error("Error writing ftdc data. Shutting down FTDC.", "err", err)
```

Comment:
The business logic for errors in the write-path. I think all of the panics I removed get to this line of code now.

Reply:
Any `writeDatum` error is fatal for FTDC?

Reply:
Oh shoot, good catch (or at least I think* you caught it).

The intention was that errors bubbling up to here are all disk related. And I'm not yet interested in making this robust across disk errors.

But one of the errors that can bubble up is the schema error. That we technically handle at a lower level (that I think is a reasonable spot). And I think bubbling it up is reasonable? But we certainly shouldn't shutdown ftdc over that case.

Reply:
I didn't really catch much, but thanks for verifying that not _all_ `writeDatum` errors are fatal to FTDC. I assume `errNotStruct` will get warn or error-logged at some point deeper in the code?

Reply:
That was the idea -- I had meant to add a test (but forgot going through the other comments). But I think I still did this wrong

Reply:
Ok, I did a thing. Not 100% I'm making the best use of error handling tools available.

But I do have a test that I can make pass/fail as I break the code. So I'm confident this does what I want.

Reply:
LGTM! Especially if test indicates you're covering "catch" path.